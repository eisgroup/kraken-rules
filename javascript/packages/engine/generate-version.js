const fs = require('fs')
const path = require('path')
const xmlParser = require('fast-xml-parser')

console.log('----------------------------------');
console.log('generating engine version constant');
console.log('----------------------------------');


const pomPath = path.join(
    __dirname,
    '../',
    '../',
    '../',
    'pom.xml'
)

console.log(`  reading '${pomPath}'`);

const pom = xmlParser.parse(
    fs.readFileSync(
        pomPath,
        { encoding: 'utf8' }
    )
)

console.log(`  version is: '${pom.project.version}'`);

const genFolderPath = path.join(
    __dirname,
    'src',
    'engine',
    'generated'
)

if (!fs.existsSync(genFolderPath)) {
    fs.mkdirSync(genFolderPath)
}

const genPath = path.join(
    __dirname,
    'src',
    'engine',
    'generated',
    'engineVersion.ts'
);

console.log(`  writing to path: '${genPath}'`);

fs.writeFileSync(
    genPath,
    `
// generated code    
// do not change it manually    
export const ENGINE_VERSION = "${pom.project.version}";

`, { encoding: 'utf8' }
)

console.log('  backend version is written');

process.exit(0)