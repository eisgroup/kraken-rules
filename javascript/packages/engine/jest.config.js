module.exports = {
    moduleFileExtensions: ['ts', 'tsx', 'js'],
    transform: {
        '^.+\\.ts$': ['@swc/jest'],
    },
    testMatch: ['**/?(*.)test.ts'],
    verbose: true,
    cacheDirectory: './target/tmp/',
    coverageDirectory: './target/coverage',
    coverageThreshold: {
        global: {
            branches: 93,
            functions: 96,
            lines: 96,
            statements: 96,
        },
    },
    collectCoverageFrom: [
        '**/*[^d\\.].ts',
        '!**/node_modules/**',
        '!**/target/**',
        '!**/__tests__/**',
        '!**/index.ts',
        '!**/**Error**.ts',
        '!**/Reducer.ts', //interface only
        '!**/RulePayloadHandler.ts', //interface only
        '!**/RepositoryClient.ts', //deprecated
        '!**/info/iterators/**', //deprecated
        '!**/config.ts',
        '!**/DevelopmentLogger.ts',
    ],
    testEnvironment: 'node',
    setupFilesAfterEnv: ['./__tests__/console.setup.js', './__tests__/matchers.setup.ts'],
}
