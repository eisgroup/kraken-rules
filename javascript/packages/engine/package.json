{
    "name": "kraken-typescript-engine",
    "version": "1.55.1-stage.0",
    "description": "",
    "main": "target/dist/js/src/index.js",
    "types": "target/dist/definitions/src/index.d.ts",
    "author": "Pavel Surinin <psurinin@eisgroup.com> (http://eisgroup.com)",
    "license": "Apache-2.0",
    "publishConfig": {
        "registry": "https://genesis-npm-release.exigengroup.com/repository/genesis-npm-release/"
    },
    "typedocMain": "src/index.ts",
    "dependencies": {
        "bignumber.js": "9.0.1",
        "declarative-js": "3.8.15",
        "fast-deep-equal": "2.0.1",
        "kraken-engine-api": "1.55.1-stage.0",
        "kraken-model": "1.55.1-stage.0",
        "moment": "2.29.4"
    },
    "devDependencies": {
        "@types/jest": "27.5.1",
        "@types/lodash": "4.14.116",
        "@types/node": "13.9.0",
        "child_process": "1.0.2",
        "fast-xml-parser": "^3.19.0",
        "jest": "27.5.1",
        "jsdom": "11.3.0",
        "jsdom-global": "3.0.2",
        "kraken-jest-matchers": "1.55.1-stage.0",
        "kraken-model-builder": "1.55.1-stage.0",
        "kraken-test-product": "1.55.1-stage.0",
        "kraken-test-product-model-tree": "1.55.1-stage.0",
        "prettier": "1.14.3",
        "rimraf": "3.0.2",
        "ts-node": "4.1.0",
        "typescript": "5.1.6"
    },
    "scripts": {
        "ci-versions": "echo \"typescript version\" && tsc -v && echo \"jest version\" && jest -v",
        "test": "jest -c jest.config.js",
        "test:coverage": "jest -c jest.config.js --coverage",
        "lint": "eslint src __tests__",
        "compile": "node generate-version && tsc --pretty",
        "postcompile": "yarn transpile",
        "transpile": "echo 'no transpile'",
        "del:compiled": "rimraf target",
        "build": "yarn run compile && yarn run lint && yarn run test:coverage",
        "prepare": "yarn run ci-versions && yarn run del:compiled && yarn run compile && yarn run lint -- --fix && yarn run test:coverage",
        "release": "yarn publish"
    },
    "files": [
        "target/dist/definitions",
        "target/dist/js"
    ],
    "gitHead": "efcf1a5e805088ee8af67ef21c2b8ce2fea62eb4"
}