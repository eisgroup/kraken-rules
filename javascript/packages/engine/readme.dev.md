# Kraken Typescript Engine

## Tests

To run tests enter command in console:
`npm test`
This command will run all tests without coverage

To run tests with coverage enter command in console:
`npm run test:coverage`

Kraken engine have its own matchers in jest, that can be used `expect(value).<matcher>`

```typescript
expect(results: EntryPointResult).k_toHaveExpressionsFailures(num?: number): void;
expect(result: RuleEvaluationResult).k_toBeValidRuleResult(): void;
expect(date: Date).k_toBeDate(argument: Date): void;
expect(date: Date).k_toBeDateTime(argument: Date): void;
expect(date: Date).k_toBeTodayDate(): void;
expect(results: EntryPointResult).k_toMatchResultsSnapshots();
/**
 * Matches rule results according to provided parameters, 
 * and check to haven no expression failures. For assert it uses
 * '.k_toHaveExpressionsFailures(0)'
 * @param stats     statistics results should match
 */
expect(results: EntryPointResult).k_toMatchResultsStats(stats: {
    total: number,
    critical?: number,
    warning?: number,
    info?: number,
    disabled?: number,
    hidden?: number
});
```

## Lint
To lint project enter command in console:
`npm run lint`

TSLint can fix some errors. To do this enter command in console:
`npm run lint -- --fix`

## Committing
Before commiting run this command:
`npm run precommit` this will fix some linter issues if they are present, run linter, 
tests with coverage and compile typescript to javascript
