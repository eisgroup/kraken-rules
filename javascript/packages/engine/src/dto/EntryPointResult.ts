/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from "./FieldEvaluationResult";
import { RuleEvaluationResults } from "./RuleEvaluationResults";
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult;

import { Reducer } from "declarative-js";
import flat = Reducer.flat;
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";

/**
 * Contains evaluation results from specific entry point
 */
export class EntryPointResult {
    private readonly results: Record<string, FieldEvaluationResult>;
    readonly evaluationTimestamp: Date;
    constructor(results: Record<string, FieldEvaluationResult>) {
        this.results = results;
        this.getApplicableResults.bind(this);
        this.evaluationTimestamp = new Date();
    }
    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     */
    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, []);
    }
    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results;
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults()
            .filter(r => ConditionEvaluationResult.isApplicable(r.conditionEvaluationResult));
    }

    static empty(): EntryPointResult {
        return new EntryPointResult({});
    }
}
