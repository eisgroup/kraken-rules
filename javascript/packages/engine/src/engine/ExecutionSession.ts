/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Rule } from 'kraken-model'
import { WithKraken } from '../config'
import { debug } from '../debugger/Debugger'
import { EvaluationConfig } from './executer/SyncEngine'
import { DataContext } from './contexts/data/DataContext'
import { DateCalculator, DefaultDateCalculator } from './runtime/expressions/date/DateCalculator'

export class ExecutionSession {
    public readonly entryPointName: string
    public readonly currencyCd: string
    public readonly expressionContext: Record<string, unknown>
    public readonly timestamp: Date
    public readonly ruleTimezoneId: string
    public readonly dateCalculator: DateCalculator
    private readonly breakPointMatcher: debug.impl.BreakPointMatcher
    constructor(
        evaluationConfig: EvaluationConfig,
        expressionContext: Record<string, unknown>,
        entryPointName: string,
    ) {
        this.currencyCd = evaluationConfig.currencyCd
        this.expressionContext = {
            external: { ...evaluationConfig.context.externalData },
            dimensions: { ...evaluationConfig.context.dimensions },
            ...expressionContext,
        }
        this.timestamp = new Date()
        this.entryPointName = entryPointName
        this.dateCalculator = evaluationConfig.dateCalculator ?? new DefaultDateCalculator()
        this.ruleTimezoneId = evaluationConfig.ruleTimezoneId ?? Intl.DateTimeFormat().resolvedOptions().timeZone
        const g: unknown = globalThis
        if (globalKrakenObjectInitialized(g)) {
            this.breakPointMatcher = new debug.impl.BreakPointMatcher(g.Kraken.debugger.breakPoints, {
                ...g.Kraken.debugger,
            })
        } else {
            this.breakPointMatcher = new debug.impl.BreakPointMatcher(new Map(), { break: false, log: false })
        }
    }

    shouldBreak(rule: Rule, dataContext: DataContext): boolean {
        return this.breakPointMatcher.ruleMatches({
            contextId: dataContext.contextId,
            contextName: dataContext.contextName,
            entryPointName: this.entryPointName,
            ruleName: rule.name,
        })
    }

    shouldBreakOnEntryPoint(): boolean {
        return this.breakPointMatcher.entryPointMatches({ entryPointName: this.entryPointName })
    }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function globalKrakenObjectInitialized(o: any): o is WithKraken {
    return 'Kraken' in o
}
