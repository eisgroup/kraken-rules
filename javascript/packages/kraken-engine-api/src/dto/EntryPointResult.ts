/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from './FieldEvaluationResult'
import { RuleEvaluationResults } from './RuleEvaluationResults'
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult

/**
 * Contains evaluation results from specific entry point
 */
export interface EntryPointResult {
    readonly evaluationTimestamp: Date
    readonly ruleTimeZoneId: string

    /**
     * Returns entry point result mapped by field name.
     */
    getFieldResults(): Record<string, FieldEvaluationResult>

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     */
    getAllRuleResults(): RuleEvaluationResult[]

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     * filtered by rule condition evaluation status.
     */
    getApplicableResults(): RuleEvaluationResult[]
}
