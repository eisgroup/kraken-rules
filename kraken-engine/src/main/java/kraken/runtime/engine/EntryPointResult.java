package kraken.runtime.engine;

import kraken.annotations.API;
import kraken.runtime.engine.dto.FieldEvaluationResult;
import kraken.runtime.engine.dto.RuleEvaluationResult;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Contains evaluation results from specific entry point
 *
 * @author rimas
 * @since 1.0
 */
@API
public class EntryPointResult {

    /**
     * Field evaluation results, mapped to field id string
     */
    private Map<String, FieldEvaluationResult> fieldResults;

    private transient LocalDateTime evaluationTimeStamp;
    /**
     * Default constructor
     */
    public EntryPointResult() {
        fieldResults = new HashMap<>();
        evaluationTimeStamp = LocalDateTime.now();
    }

    public EntryPointResult(Map<String, FieldEvaluationResult> fieldResults, LocalDateTime evaluationTimeStamp) {
        this.fieldResults = new HashMap<>(fieldResults);
        this.evaluationTimeStamp = evaluationTimeStamp;
    }

    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     *
     * @return  list of all contained {@link RuleEvaluationResult}s
     */
    public List<RuleEvaluationResult> getAllRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .collect(Collectors.toList());
    }

    public List<RuleEvaluationResult> getApplicableRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .filter(r -> r.getConditionEvaluationResult().isApplicable())
                .collect(Collectors.toList());
    }

    public LocalDateTime getEvaluationTimeStamp() {
        return evaluationTimeStamp;
    }

    public Map<String, FieldEvaluationResult> getFieldResults() {
        return fieldResults;
    }

}
