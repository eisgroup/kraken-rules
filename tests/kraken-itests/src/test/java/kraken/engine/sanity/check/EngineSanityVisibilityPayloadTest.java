/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.engine.sanity.check;

import kraken.runtime.engine.EntryPointResult;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static io.github.jsonSnapshot.SnapshotMatcher.validateSnapshots;
import static kraken.testing.matchers.KrakenMatchers.*;
import static kraken.test.KrakenItestMatchers.matchesSnapshot;
import static org.junit.Assert.assertThat;

public final class EngineSanityVisibilityPayloadTest extends SanityEngineBaseTest {

    @BeforeClass
    public static void beforeAll() {
        start();
    }

    @AfterClass
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    public void shouldExecuteVisibilityAutoPolicyEntryPoint() {
        final EntryPointResult result = engine.evaluate(dataObject, "VisibilityAutoPolicy");

        assertThat(result, matchesSnapshot());
        assertThat(result, hasHiddenFields(7));
        assertThat(result, hasRuleResults(7));
    }
}
