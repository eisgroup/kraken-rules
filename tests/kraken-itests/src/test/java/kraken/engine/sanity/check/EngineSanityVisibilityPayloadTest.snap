kraken.engine.sanity.check.EngineSanityVisibilityPayloadTest.shouldExecuteVisibilityAutoPolicyEntryPoint=[
  {
    "Policy:-1:createdBy": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "createdBy",
        "fieldPath": "accessTrackInfo.createdBy"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "R0034",
            "context": "Policy",
            "targetPath": "createdBy",
            "payloadType": "VISIBILITY"
          },
          "payloadResult": {
            "visible": false
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:createdFromPolicyRev": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "createdFromPolicyRev",
        "fieldPath": "createdFromPolicyRev"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "R0010",
            "context": "Policy",
            "targetPath": "createdFromPolicyRev",
            "payloadType": "VISIBILITY"
          },
          "payloadResult": {
            "visible": false
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:createdOn": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "createdOn",
        "fieldPath": "accessTrackInfo.createdOn"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "R0030",
            "context": "Policy",
            "targetPath": "createdOn",
            "payloadType": "VISIBILITY"
          },
          "payloadResult": {
            "visible": false
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:state": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "state",
        "fieldPath": "state"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "R0006",
            "context": "Policy",
            "targetPath": "state",
            "payloadType": "VISIBILITY"
          },
          "payloadResult": {
            "visible": false
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:txType": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "txType",
        "fieldPath": "transactionDetails.txType"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "R0012",
            "context": "Policy",
            "targetPath": "txType",
            "payloadType": "VISIBILITY"
          },
          "payloadResult": {
            "visible": false
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:updatedBy": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "updatedBy",
        "fieldPath": "accessTrackInfo.updatedBy"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "R0041",
            "context": "Policy",
            "targetPath": "updatedBy",
            "payloadType": "VISIBILITY"
          },
          "payloadResult": {
            "visible": false
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    },
    "Policy:-1:updatedOn": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "updatedOn",
        "fieldPath": "accessTrackInfo.updatedOn"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "R0038",
            "context": "Policy",
            "targetPath": "updatedOn",
            "payloadType": "VISIBILITY"
          },
          "payloadResult": {
            "visible": false
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]